File: frontend/client/src/types/Messages.ts
import { UUID } from '@/types/GenericTypes';
import { CardDto } from '@/models/CardDto';
import { RoundState } from '@/types/RoundState';

interface CommonMessage {
  id: UUID;
  type: string;
  payload: object;
}

export interface RefreshMessage extends CommonMessage {
  type: 'refresh';
  payload: {
    card: CardDto;
  };
}

export interface RoundMessage extends CommonMessage {
  type: 'round';
  payload: {
    isWin: boolean;
    card: CardDto;
    round_state?: RoundState;
  };
}

export interface CounterMessage extends CommonMessage {
  type: 'counter';
  payload: {
    counter: string;
  };
}

export interface UpdateGameStateMessage extends CommonMessage {
  type: 'state';
  payload: {
    counter?: string;
    cards: CardDto[];
    round_state?: RoundState;
    current_round?: number;
  };
}

export type CommunicationMessages = RoundMessage | RefreshMessage | CounterMessage | UpdateGameStateMessage;

File: backend/dealer/messages.go
package main

const (
	MessageTypeFullState = "state"
	MessageTypeCounter   = "counter"
	MessageTypeRound     = "round"
	MessageTypeCardState = "card_state"
)

// Message object
type Message struct {
	Type    string      `json:"type"`
	Payload interface{} `json:"payload"`
}

// StateMessagePayload a struct for payload
type StateMessagePayload struct {
	Cards        []Card `json:"cards,omitempty"`
	RoundState   string `json:"round_state,omitempty"`
	Counter      string `json:"counter,omitempty"`
	CurrentRound uint8  `json:"current_round,omitempty"`
}

// MessageCounterPayload struct
type MessageCounterPayload struct {
	Counter string `json:"counter,omitempty"`
}

type MessageCardStatePayload struct {
	Card  Card `json:"card,omitempty"`
	RoundState   string `json:"round_state,omitempty"`
	Counter      string `json:"counter,omitempty"`
	CurrentRound uint8  `json:"current_round,omitempty"`
}

// RoundMessagePayload is the payload
type RoundMessagePayload struct {
	IsWin bool `json:"is_win"`
	Card  Card `json:"card"`
}
